File: wow-core/src/main/kotlin/me/ahoo/wow/event/DomainEventStream.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.ahoo.wow.event

import me.ahoo.wow.api.Copyable
import me.ahoo.wow.api.Version
import me.ahoo.wow.api.command.RequestId
import me.ahoo.wow.api.event.DomainEvent
import me.ahoo.wow.api.event.EventMessage
import me.ahoo.wow.api.event.IgnoreSourcing
import me.ahoo.wow.api.exception.ErrorInfo
import me.ahoo.wow.api.messaging.Header
import me.ahoo.wow.api.messaging.NamedBoundedContextMessage
import me.ahoo.wow.api.modeling.AggregateId
import me.ahoo.wow.api.modeling.AggregateIdCapable
import me.ahoo.wow.api.modeling.SpaceId
import me.ahoo.wow.id.generateGlobalId
import me.ahoo.wow.messaging.DefaultHeader

/**
 * Domain Event Stream interface representing a sequence of domain events.
 *
 * A domain event stream contains a collection of domain events that were generated
 * as a result of a single command execution. The relationship is 1:1 between
 * event streams and command IDs.
 *
 * Key requirements:
 * - Events must be sorted in ascending order by version number
 * - Version numbers must be monotonically increasing
 * - All events in a stream belong to the same aggregate
 * - Events are immutable once created
 *
 * @property aggregateId The aggregate ID this event stream belongs to
 * @property size The number of events in this stream
 *
 * @see DomainEvent
 * @see NamedBoundedContextMessage
 * @see AggregateIdCapable
 * @see Copyable
 */
interface DomainEventStream :
    EventMessage<DomainEventStream, List<DomainEvent<*>>>,
    RequestId,
    Iterable<DomainEvent<*>>,
    Copyable<DomainEventStream> {
    override val aggregateId: AggregateId
    val size: Int
}

/**
 * Simple implementation of DomainEventStream.
 *
 * This data class provides a concrete implementation of the DomainEventStream interface,
 * containing a list of domain events with associated metadata.
 *
 * @property id The unique identifier of this event stream (default: generated global ID)
 * @property requestId The request ID that initiated this event stream
 * @property header The message header containing metadata (default: empty header)
 * @property body The list of domain events in this stream
 * @property aggregateId The aggregate ID (derived from the first event)
 * @property contextName The bounded context name (derived from aggregateId)
 * @property aggregateName The aggregate name (derived from aggregateId)
 * @property ownerId The owner ID (derived from the first event)
 * @property commandId The command ID (derived from the first event)
 * @property version The aggregate version (derived from the first event)
 * @property size The number of events in the stream
 * @property createTime The creation timestamp (derived from the first event)
 *
 * @constructor Creates a new SimpleDomainEventStream
 * @param id The stream ID
 * @param requestId The request ID
 * @param header The message header
 * @param body The list of domain events (must not be empty)
 * @throws IllegalArgumentException if the event list is empty
 *
 * @see DomainEventStream
 * @see DomainEvent
 * @see Header
 */
data class SimpleDomainEventStream(
    override val id: String = generateGlobalId(),
    override val requestId: String,
    override val header: Header = DefaultHeader.empty(),
    override val body: List<DomainEvent<*>>
) : DomainEventStream,
    Iterable<DomainEvent<*>> by body {
    override val aggregateId: AggregateId

    override val contextName: String
        get() = aggregateId.contextName
    override val aggregateName: String
        get() = aggregateId.aggregateName
    override val ownerId: String
    override val spaceId: SpaceId
    override val commandId: String
    override val version: Int

    override fun copy(): DomainEventStream = copy(header = header.copy())

    override val size: Int
    override val createTime: Long

    init {
        require(body.isNotEmpty()) { "events can not be empty." }
        body.first().let {
            aggregateId = it.aggregateId
            ownerId = it.ownerId
            spaceId = it.spaceId
            commandId = it.commandId
            version = it.version
            createTime = it.createTime
        }
        size = body.size
    }
}

/**
 * Determines if this event stream should be ignored during event sourcing.
 *
 * This function checks if the event stream contains only events that should be
 * ignored during the event sourcing process. An event stream is considered
 * ignorable if it represents the initial version and all events in the stream
 * are marked with IgnoreSourcing and contain ErrorInfo.
 *
 * @return true if the event stream should be ignored during sourcing, false otherwise
 *
 * @see IgnoreSourcing
 * @see ErrorInfo
 * @see Version.isInitialVersion
 */
fun DomainEventStream.ignoreSourcing(): Boolean {
    if (!isInitialVersion) {
        return false
    }
    return body.all {
        it.body is IgnoreSourcing && it.body is ErrorInfo
    }
}

File: typescript/wow-client/src/query/event/domainEventStream.ts
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  AggregateId,
  CreateTimeCapable,
  Identifier,
  Named,
  OwnerId,
  Version,
} from '../../types';
import { CommandId, CommandStage, RequestId } from '../../command';
import { BodyCapable } from '../../types';
import { JsonServerSentEvent } from '@ahoo-wang/fetcher-eventstream';

/**
 * Represents a domain event with a specific body type.
 * Extends Identifier, Named, and BodyCapable interfaces to provide identification,
 * naming, and body capabilities for the domain event.
 * @template BODY - The type of the event content for this domain event
 */
export interface DomainEvent<BODY>
  extends Identifier,
    Named,
    BodyCapable<BODY> {
  /**
   * The type of the event content.
   */
  bodyType: string;
  /**
   * The revision of the domain event.
   */
  revision: string;
}

/**
 * Represents the header information for a domain event stream.
 * Contains metadata about the event stream such as command information,
 * network details, and tracing information.
 */
export interface DomainEventStreamHeader {
  /**
   * The operator that executed the command.
   */
  command_operator?: string;
  /**
   * The endpoint to wait for command completion.
   */
  command_wait_endpoint?: string;
  /**
   * The stage to wait for in command execution.
   */
  command_wait_stage?: CommandStage;

  local_first?: string;
  /**
   * The IP address of the remote client.
   */
  remote_ip: string;
  /**
   * The user agent of the client.
   */
  user_agent?: string;
  /**
   * The trace identifier for distributed tracing.
   */
  trace_id?: string;

  /**
   * Index signature for additional custom header properties.
   * Allows for any additional string key-value pairs to be included as header properties.
   */
  [key: string]: string | undefined;
}

/**
 * Represents a stream of domain events.
 * Combines multiple interfaces to provide a complete domain event stream,
 * including identification, aggregation, ownership, command information,
 * versioning, and the actual event data.
 */
export interface DomainEventStream
  extends Identifier,
    AggregateId,
    OwnerId,
    CommandId,
    CreateTimeCapable,
    RequestId,
    Version,
    BodyCapable<DomainEvent<any>[]> {
  /**
   * The header information for the domain event stream.
   */
  header: DomainEventStreamHeader;
}

/**
 * Provides field names for domain event stream metadata.
 *
 * This class contains static readonly properties that define the field names used in domain event stream metadata.
 * These field names are used to access and manipulate domain event stream data in a consistent manner.
 * The fields include headers, identifiers, command information, versioning, body content, and creation time.
 */
export class DomainEventStreamMetadataFields {
  static readonly HEADER = 'header';
  static readonly COMMAND_OPERATOR = `${DomainEventStreamMetadataFields.HEADER}.command_operator`;
  static readonly AGGREGATE_ID = 'aggregateId';
  static readonly TENANT_ID = 'tenantId';
  static readonly OWNER_ID = 'ownerId';
  static readonly COMMAND_ID = 'commandId';
  static readonly REQUEST_ID = 'requestId';
  static readonly VERSION = 'version';
  static readonly BODY = 'body';
  static readonly BODY_ID = `${DomainEventStreamMetadataFields.BODY}.id`;
  static readonly BODY_NAME = `${DomainEventStreamMetadataFields.BODY}.name`;
  static readonly BODY_TYPE = `${DomainEventStreamMetadataFields.BODY}.bodyType`;
  static readonly BODY_REVISION = `${DomainEventStreamMetadataFields.BODY}.revision`;
  static readonly BODY_BODY = `${DomainEventStreamMetadataFields.BODY}.body`;
  static readonly CREATE_TIME = 'createTime';
}

/**
 * Represents a readable stream of domain event streams.
 *
 * This type defines a ReadableStream that emits JsonServerSentEvent objects containing DomainEventStream data.
 * It is used for streaming domain events in a server-sent event format.
 */
export type ReadableDomainEventStream = ReadableStream<
  JsonServerSentEvent<DomainEventStream>
>;
